### DISCUSSION TITLE
PyScript terminal support stdin using a static standalone file?

### DISCUSSION THREAD

**Discussion by engdan77** [pyscript/pyscript#2102] [Q&A] [answered]
Now if we could look at this example below to easier describe the use case.

```python
<html>
<head>
<script type="module" src="https://pyscript.net/releases/2024.6.1/core.js"></script>
<script scope="./">
/*! coi-serviceworker v0.1.7 - Guido Zuidhof and contributors, licensed under MIT */
/*! mini-coi - Andrea Giammarchi and contributors, licensed under MIT */
(({ document: d, navigator: { serviceWorker: s } }) => {
  if (d) {
    const { currentScript: c } = d;
    s.register(c.src, { scope: c.getAttribute('scope') || '.' }).then(r => {
      r.addEventListener('updatefound', () => location.reload());
      if (r.active && !s.controller) location.reload();
    });
  }
  else {
    addEventListener('install', () => skipWaiting());
    addEventListener('activate', e => e.waitUntil(clients.claim()));
    addEventListener('fetch', e => {
      const { request: r } = e;
      if (r.cache === 'only-if-cached' && r.mode !== 'same-origin') return;
      e.respondWith(fetch(r).then(r => {
        const { body, status, statusText } = r;
        if (!status || status > 399) return r;
        const h = new Headers(r.headers);
        h.set('Cross-Origin-Opener-Policy', 'same-origin');
        h.set('Cross-Origin-Embedder-Policy', 'require-corp');
        h.set('Cross-Origin-Resource-Policy', 'cross-origin');
        return new Response(body, { status, statusText, headers: h });
      }));
    });
  }
})(self);
</script>
</head>
<body>

<script type="py" async terminal worker>
    print('Hello!!')
    name = input('What is your name?')
    print(f'Welcome {name}')    
</script>
</body>
</html>
```
_We have embedded COI to ensure worker to work without serving this using web service._

Imagine you would like to share such example by sharing this HTML file being "offline" CLI based tool that allow input/output to someone <ins>without need open it within a a web server</ins> to allow them to use this.

Now the problem with above is if you would open this in your browser (e.g. Chrome) you'll get an error that appears as:
```
error: TypeError: "file:///Users/edo/tmp/pyscript/config.txt" cannot be parsed as a URL against "null/0b381238-00f3-4a11-b4e2-5c7656d96ed3".
```

Buf it you host this using e.g.
```
python3 -m http.server
```

It will run fine at least using `<script src="./mini-coi.js" scope="./">`
Do you think this is a bug? 
_... or is it a technical limitation that "workers" are not allowed to start opening the page as a static file (e.g. opening as a file on your local storage) so we cannot expect PyScript's terminal to function with stdin anytime in a near future?_

...

The only current workaround I've seen suggested would be to re-implement rather using the JS prompt instead
```python
import contextlib

with contextlib.suppress(ImportError):
    from pyscript import window
    input = window.prompt

name1 = input('name?')
```
https://stackoverflow.com/questions/76001015/how-do-you-use-an-input-in-pyscript

**Comment by WebReflection** [comment]
> We have embedded COI to ensure worker to work without serving this using web service.

You **need** to have that `mini-coi.js` **file**, not as inline script as there's no way to register it as Service Worker if it's inline.

[Documentation](https://docs.pyscript.net/2024.6.1/user-guide/workers/#http-headers)
[mini-coi](https://github.com/WebReflection/mini-coi?tab=readme-ov-file#how-to-use-mini-coi-as-service-worker)

Let's try again:

```html
<!doctype html>
<html>
<head>
    <script src="mini-coi.js"></script>
    <script type="module" src="https://pyscript.net/releases/2024.6.1/core.js"></script>
</head>
<body>
    <script type="py" async terminal worker>
        print('Hello!!')
        name = input('What is your name?')
        print(f'Welcome {name}')    
    </script>
</body>
</html>
```

Now reach `http://localhost:8000/` (not `0.0.0.0:8080`, it has to be localhost or no ServiceWorker is available because insecure without `https`) and see it working.

**Reply by engdan77** [reply]
Thanks a whole lot @WebReflection that's really good to know, and will keep this in mind.
... however, the main question should probably be framed differently that I am curious whether you or anyone else might have the answer to (possibly good addition to the "terminal" section in our documentation):

> Is it possible (at all, using any techniques) to use PyScript including a terminal supporting input solely opening the HTML as a static local file on you storage without any need of a web server?

**Reply by WebReflection** [reply]
with MicroPython yes, try this instead:

```html
<!doctype html>
<html>
<head>
    <script src="mini-coi.js"></script>
    <script type="module" src="https://pyscript.net/releases/2024.6.1/core.js"></script>
</head>
<body>
    <script type="mpy" async terminal>
        print('Hello!!')
        name = input('What is your name?')
        print(f'Welcome {name}')    
    </script>
</body>
</html>
```

however, that's not *worker* based because workers out of a file system and without SharedArrayBuffer can't guarantee:

  * that input is not blocking, you will indeed see a prompt in this *mpy* example, we can't block the main, that's by contract on the Web
  * if you use `time.sleep(1)` you block the main and you won't see anything, as exampe in a while loop, until the whole loop has finished ... that's also bad for the browser which might complain something is blocking and asking you to kill the tab

If we are talking about *Electron* like applications though, or *Tauri* based apps, most browsers have flags to allow files from the system as secure, force-enable SharedArrayBuffer and all other sort of security risks that are mitigated on the Web so you might be lucky there but I don't know if that's what you are after.

**Reply by WebReflection** [reply]
P.S. frgot to mention you can also use all these features within an intranet and without issues, see flags passed along to start chrome/ium [here](https://github.com/WebReflection/coincident-oled?tab=readme-ov-file#launch-chromeium)

**Reply by engdan77** [reply]
Thanks again @WebReflection, the solution that MicroPython `name = input("Your name: ")` looks similar previously mentioned `name = windows.prompt("Your name: ")` is perhaps less elegant as such an inline/interactive prompt within the terminal. However, I think this is better than not supporting input at all.  An example of those differences is found in the attached video.

I'll also check closer into the `--unsafely-treat-insecure-origin-as-secure=http://...` however presume also need an HTTP server to be of value.

I guess how I currently look at this is that for a true "interactive" terminal we might need to lean towards using "worker", and for a worker to function we'd still require an HTTP server (either serving mini-coi.js or those proper headers) - if I've understood things right? :-) 

![example](https://github.com/pyscript/pyscript/assets/8431437/af0d9831-a148-44f6-9240-36a539fa5695)

**Reply by WebReflection** [reply, marked as answer]
the use case for serverless "*Web Apps*" faded in time as more and more APIs actually exist only in secure content ... but then again, if you can control the browser/chrome bootstrap you have options.

```sh
chromium --allow-running-insecure-content --allow-file-access-from-files --enable-features=SharedArrayBuffer ./index.html
```

With this line and this content in the `index.html` file I can interact with the terminal without needing anything special at all:

```html
<!doctype html>
<html>
<head>
    <script type="module" src="https://pyscript.net/releases/2024.6.1/core.js"></script>
</head>
<body>
    <script id="mpy-terminal" type="mpy" worker terminal>
        import code
        code.interact()
    </script>
</body>
</html>
```

So ... there is a way to allow special things in browsers but every browser might have different flags configuration although with the `chromium` one, which is behind Edge, Chrome, and others, you should be able to move forward?

**Reply by engdan77** [reply]
@WebReflection great tip of advice..  🥇   

Thanks to this it sure does work run this as a static "serverless" PyScript including a Terminal input such as this example (based on yours)

```html
<!doctype html>
<html>
<head>
    <script type="module" src="https://pyscript.net/releases/2024.6.1/core.js"></script>
</head>
<body>
    <script id="py-terminal" type="py" worker terminal>
        import code
        x = input('What is your name: ')
        print(f'Hello {x}')
    </script>
</body>
</html>
```


Useful under circumstances where you wish to share an "interactive" (a simple terminal-based) with someone who has no access to the Internet and nothing else installed on the computer but Chrome - even though with some arguments needed for Chrome but less of a problem.

On macOS, you would start Chrome:

```
open -a "Google Chrome" --args --new --allow-running-insecure-content --allow-file-access-from-files --enable-feat
ures=SharedArrayBuffer ./index.html
``` 

And that would allow you to interact through this terminal.

Thanks again, and hopefully this is a _recipe_ that might help others asking for the same. 🥳 👍🏻

**Reply by WebReflection** [reply]
yeah, this probably can go into our *F.A.Q.* section /cc @ntoll
